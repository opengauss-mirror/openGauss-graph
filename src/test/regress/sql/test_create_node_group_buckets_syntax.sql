DECLARE
	group_count              int;
	i                        int;
	node_names               text;
BEGIN
	select count(*) from pgxc_group into group_count;
	IF group_count != 1
	THEN
		RAISE INFO 'current group_count is %', group_count;
		i := 1;
		FOR group_p in (select * from pgxc_group)
		LOOP
			select string_agg(node_name,',') from pgxc_node, pgxc_group where pgxc_node.oid = any(group_members) and pgxc_group.oid = group_p.oid into node_names;
			RAISE INFO 'group% name: %, members: %', i, group_p.group_name, node_names;
			i := i+1;
		END LOOP;
		RETURN;
	END IF;
	
	RAISE INFO 'check group_count ok.';
	RETURN;
END;
/

DECLARE
	group_info               record;
	node_names               text;
	query                    text;
	test_case                text;
BEGIN
	test_case := 'create group with buckets';
	select * from pgxc_group into group_info;
	select string_agg(node_name,',') from pgxc_node, pgxc_group where pgxc_node.oid = any(group_members) and pgxc_group.oid = group_info.oid into node_names;
	
	RAISE INFO 'test %...', test_case;
	query := 'create node group ' || group_info.group_name || '_dummy with(' || node_names || ') buckets(' || group_info.group_buckets || ')';
	execute query;
	
	query := 'drop node group ' || group_info.group_name || '_dummy';
	execute query;
	RAISE INFO 'test % ok', test_case;
	RETURN;
END;
/

DECLARE
	group_info               record;
	node_names               text;
	query                    text;
	test_case                text;
BEGIN
	test_case := 'create group with buckets number not correct';
	select * from pgxc_group into group_info;
	select string_agg(node_name,',') from pgxc_node, pgxc_group where pgxc_node.oid = any(group_members) and pgxc_group.oid = group_info.oid into node_names;
	
	RAISE INFO 'test %...', test_case;
	query := 'create node group ' || group_info.group_name || '_dummy with(' || node_names || ') buckets(' || group_info.group_buckets || ',0)';
	execute query;
	
	RAISE INFO 'test % ok', test_case;
	RETURN;
END;
/

DECLARE
	group_info               record;
	node_names               text;
	bucket_str               text;
	query                    text;
	test_case                text;
BEGIN
	test_case := 'create group with buckets not even: some node has more buckets';
	select * from pgxc_group into group_info;
	select string_agg(node_name,',') from pgxc_node, pgxc_group where pgxc_node.oid = any(group_members) and pgxc_group.oid = group_info.oid into node_names;
	
	RAISE INFO 'test %...', test_case;
	select regexp_replace(group_info.group_buckets,'(^|,)0($|,)','\11\2') into bucket_str;
	select regexp_replace(bucket_str,'(^|,)0($|,)','\11\2') into bucket_str;
	select regexp_replace(bucket_str,'(^|,)0($|,)','\11\2') into bucket_str;
	query := 'create node group ' || group_info.group_name || '_dummy with(' || node_names || ') buckets(' || bucket_str || ')';
	execute query;
	
	RAISE INFO 'test % ok', test_case;
	RETURN;
END;
/

DECLARE
	group_info               record;
	node_names               text;
	bucket_str               text;
	query                    text;
	test_case                text;
BEGIN
	test_case := 'create group with buckets not even: some node has no buckets';
	select * from pgxc_group into group_info;
	select string_agg(node_name,',') from pgxc_node, pgxc_group where pgxc_node.oid = any(group_members) and pgxc_group.oid = group_info.oid into node_names;
	
	RAISE INFO 'test %...', test_case;
	select regexp_replace(group_info.group_buckets,'(^|,)0($|,)','\11\2','g') into bucket_str;
	query := 'create node group ' || group_info.group_name || '_dummy with(' || node_names || ') buckets(' || bucket_str || ')';
	execute query;
	
	RAISE INFO 'test % ok', test_case;
	RETURN;
END;
/

DECLARE
	node_count               int;
	group_info               record;
	node_names               text;
	bucket_str               text;
	query                    text;
	test_case                text;
BEGIN
	test_case := 'create group with bucket id out of range';
	select * from pgxc_group into group_info;
	select string_agg(node_name,',') from pgxc_node, pgxc_group where pgxc_node.oid = any(group_members) and pgxc_group.oid = group_info.oid into node_names;
	
	RAISE INFO 'test %...', test_case;
	select count(*) from pgxc_node, pgxc_group where pgxc_node.oid = any(group_members) and pgxc_group.oid = group_info.oid into node_count;
	select regexp_replace(group_info.group_buckets,'(^|,)0($|,)','\1'||node_count||'\2') into bucket_str;
	query := 'create node group ' || group_info.group_name || '_dummy with(' || node_names || ') buckets(' || bucket_str || ')';
	execute query;
	
	RAISE INFO 'test % ok', test_case;
	RETURN;
END;
/
