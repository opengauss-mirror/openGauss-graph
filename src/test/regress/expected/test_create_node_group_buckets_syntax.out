DECLARE
	group_count              int;
	i                        int;
	node_names               text;
BEGIN
	select count(*) from pgxc_group into group_count;
	IF group_count != 1
	THEN
		RAISE INFO 'current group_count is %', group_count;
		i := 1;
		FOR group_p in (select * from pgxc_group)
		LOOP
			select string_agg(node_name,',') from pgxc_node, pgxc_group where pgxc_node.oid = any(group_members) and pgxc_group.oid = group_p.oid into node_names;
			RAISE INFO 'group% name: %, members: %', i, group_p.group_name, node_names;
			i := i+1;
		END LOOP;
		RETURN;
	END IF;
	
	RAISE INFO 'check group_count ok.';
	RETURN;
END;
/
INFO:  check group_count ok.
DECLARE
	group_info               record;
	node_names               text;
	query                    text;
	test_case                text;
BEGIN
	test_case := 'create group with buckets';
	select * from pgxc_group into group_info;
	select string_agg(node_name,',') from pgxc_node, pgxc_group where pgxc_node.oid = any(group_members) and pgxc_group.oid = group_info.oid into node_names;
	
	RAISE INFO 'test %...', test_case;
	query := 'create node group ' || group_info.group_name || '_dummy with(' || node_names || ') buckets(' || group_info.group_buckets || ')';
	execute query;
	
	query := 'drop node group ' || group_info.group_name || '_dummy';
	execute query;
	RAISE INFO 'test % ok', test_case;
	RETURN;
END;
/
INFO:  test create group with buckets...
INFO:  test create group with buckets ok
DECLARE
	group_info               record;
	node_names               text;
	query                    text;
	test_case                text;
BEGIN
	test_case := 'create group with buckets number not correct';
	select * from pgxc_group into group_info;
	select string_agg(node_name,',') from pgxc_node, pgxc_group where pgxc_node.oid = any(group_members) and pgxc_group.oid = group_info.oid into node_names;
	
	RAISE INFO 'test %...', test_case;
	query := 'create node group ' || group_info.group_name || '_dummy with(' || node_names || ') buckets(' || group_info.group_buckets || ',0)';
	execute query;
	
	RAISE INFO 'test % ok', test_case;
	RETURN;
END;
/
INFO:  test create group with buckets number not correct...
ERROR:  The buckets number(1025) is not correct.
CONTEXT:  SQL statement "create node group new_group_004_dummy with(datanode1,datanode2,datanode3,datanode4,datanode5,datanode6,datanode7,datanode8,datanode9,datanode10,datanode11,datanode12,datanode13,datanode14,datanode15,datanode16) buckets(14,15,12,14,14,15,8,8,15,13,14,12,13,14,15,9,9,9,9,9,14,9,15,10,15,11,10,10,10,10,10,10,11,11,11,11,11,11,11,11,12,12,11,12,12,12,12,12,13,13,12,12,13,13,13,13,14,14,13,13,13,13,14,14,15,15,14,14,14,14,15,15,0,1,15,15,15,15,15,15,8,8,8,8,10,8,11,8,0,1,2,3,4,5,6,7,13,9,9,9,9,12,9,13,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,10,10,2,3,4,5,6,7,0,1,12,10,14,10,10,10,11,11,2,3,4,5,6,7,0,1,11,14,11,15,11,11,8,8,12,8,13,8,8,8,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,9,9,14,9,14,11,9,15,0,1,2,3,4,5,6,7,14,12,2,3,15,12,12,12,0,1,12,12,4,5,6,7,15,13,2,3,13,13,13,13,0,1,13,13,4,5,6,7,0,1,2,3,4,5,6,7,10,14,8,11,15,8,8,8,15,10,2,3,4,5,6,7,0,1,10,10,10,10,12,10,14,14,2,3,4,5,14,14,9,9,9,9,9,12,10,9,11,12,14,14,14,14,6,7,0,1,11,11,13,11,11,11,15,15,2,3,4,5,15,15,0,1,15,15,15,15,6,7,0,1,2,3,4,5,6,7,8,8,12,8,13,8,13,8,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,12,12,2,3,12,12,14,12,9,13,11,9,9,9,9,9,10,10,12,15,4,5,6,7,0,1,10,10,10,10,14,10,0,1,2,3,4,5,6,7,13,13,2,3,14,13,13,15,11,8,8,8,8,8,14,8,11,14,13,13,4,5,6,7,0,1,11,11,15,12,11,11,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,9,14,10,9,9,15,12,11,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,14,14,2,3,4,5,14,14,15,8,15,12,8,13,8,8,12,10,14,14,14,14,6,7,0,1,10,10,13,10,10,10,0,1,2,3,4,5,6,7,15,15,2,3,4,5,15,15,12,12,15,15,12,12,12,12,11,11,12,12,15,15,6,7,9,9,9,13,9,9,10,9,0,1,11,11,11,11,11,11,0,1,2,3,4,5,6,7,13,8,8,8,10,8,8,8,0,1,2,3,4,5,6,7,13,13,2,3,13,13,13,13,0,1,14,13,4,5,6,7,0,1,2,3,4,5,6,7,10,10,2,3,4,5,6,7,0,1,10,10,14,14,10,10,0,1,2,3,4,5,6,7,9,9,9,9,11,9,9,12,0,1,2,3,4,5,6,7,8,8,12,8,12,14,8,8,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,14,11,2,3,4,5,6,7,0,1,11,11,13,15,11,11,0,1,2,3,4,5,6,7,15,12,2,3,12,12,12,12,14,14,12,12,4,5,14,14,0,1,14,14,14,14,6,7,14,9,10,9,9,13,15,9,15,8,13,15,8,8,8,11,15,10,2,3,4,5,6,7,0,1,10,10,10,10,10,10,0,1,2,3,4,5,6,7,13,13,2,3,13,13,13,13,15,15,13,13,4,5,15,15,0,1,15,15,15,15,6,7,0,1,2,3,4,5,6,7,11,11,2,3,4,5,6,7,0,1,11,11,12,11,11,11,8,8,8,8,12,12,8,8,9,11,9,9,9,9,10,9,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,10,10,2,3,4,5,6,7,0,1,10,14,13,10,10,10,12,14,2,3,12,12,12,12,0,1,12,12,4,5,6,7,0,1,2,3,4,5,6,7,8,8,14,13,14,8,8,8,0,1,2,3,4,5,6,7,12,9,9,9,9,14,11,14,0,1,2,3,4,5,6,7,14,14,2,3,4,5,6,7,11,11,2,3,13,13,13,13,13,13,11,11,11,11,11,12,0,1,13,13,4,5,6,7,0,1,2,3,4,5,6,7,10,10,2,3,4,5,6,7,10,8,8,8,8,8,10,10,0,1,2,3,4,5,15,15,0,1,15,15,15,15,9,9,9,9,9,9,9,9,6,7,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,0)"
PL/pgSQL function inline_code_block line 12 at EXECUTE statement
DECLARE
	group_info               record;
	node_names               text;
	bucket_str               text;
	query                    text;
	test_case                text;
BEGIN
	test_case := 'create group with buckets not even: some node has more buckets';
	select * from pgxc_group into group_info;
	select string_agg(node_name,',') from pgxc_node, pgxc_group where pgxc_node.oid = any(group_members) and pgxc_group.oid = group_info.oid into node_names;
	
	RAISE INFO 'test %...', test_case;
	select regexp_replace(group_info.group_buckets,'(^|,)0($|,)','\11\2') into bucket_str;
	select regexp_replace(bucket_str,'(^|,)0($|,)','\11\2') into bucket_str;
	select regexp_replace(bucket_str,'(^|,)0($|,)','\11\2') into bucket_str;
	query := 'create node group ' || group_info.group_name || '_dummy with(' || node_names || ') buckets(' || bucket_str || ')';
	execute query;
	
	RAISE INFO 'test % ok', test_case;
	RETURN;
END;
/
INFO:  test create group with buckets not even: some node has more buckets...
ERROR:  Buckets distribution is not even(max_buckets: 67, min_buckets: 61).
CONTEXT:  SQL statement "create node group new_group_004_dummy with(datanode1,datanode2,datanode3,datanode4,datanode5,datanode6,datanode7,datanode8,datanode9,datanode10,datanode11,datanode12,datanode13,datanode14,datanode15,datanode16) buckets(14,15,12,14,14,15,8,8,15,13,14,12,13,14,15,9,9,9,9,9,14,9,15,10,15,11,10,10,10,10,10,10,11,11,11,11,11,11,11,11,12,12,11,12,12,12,12,12,13,13,12,12,13,13,13,13,14,14,13,13,13,13,14,14,15,15,14,14,14,14,15,15,1,1,15,15,15,15,15,15,8,8,8,8,10,8,11,8,1,1,2,3,4,5,6,7,13,9,9,9,9,12,9,13,1,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,10,10,2,3,4,5,6,7,0,1,12,10,14,10,10,10,11,11,2,3,4,5,6,7,0,1,11,14,11,15,11,11,8,8,12,8,13,8,8,8,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,9,9,14,9,14,11,9,15,0,1,2,3,4,5,6,7,14,12,2,3,15,12,12,12,0,1,12,12,4,5,6,7,15,13,2,3,13,13,13,13,0,1,13,13,4,5,6,7,0,1,2,3,4,5,6,7,10,14,8,11,15,8,8,8,15,10,2,3,4,5,6,7,0,1,10,10,10,10,12,10,14,14,2,3,4,5,14,14,9,9,9,9,9,12,10,9,11,12,14,14,14,14,6,7,0,1,11,11,13,11,11,11,15,15,2,3,4,5,15,15,0,1,15,15,15,15,6,7,0,1,2,3,4,5,6,7,8,8,12,8,13,8,13,8,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,12,12,2,3,12,12,14,12,9,13,11,9,9,9,9,9,10,10,12,15,4,5,6,7,0,1,10,10,10,10,14,10,0,1,2,3,4,5,6,7,13,13,2,3,14,13,13,15,11,8,8,8,8,8,14,8,11,14,13,13,4,5,6,7,0,1,11,11,15,12,11,11,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,9,14,10,9,9,15,12,11,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,14,14,2,3,4,5,14,14,15,8,15,12,8,13,8,8,12,10,14,14,14,14,6,7,0,1,10,10,13,10,10,10,0,1,2,3,4,5,6,7,15,15,2,3,4,5,15,15,12,12,15,15,12,12,12,12,11,11,12,12,15,15,6,7,9,9,9,13,9,9,10,9,0,1,11,11,11,11,11,11,0,1,2,3,4,5,6,7,13,8,8,8,10,8,8,8,0,1,2,3,4,5,6,7,13,13,2,3,13,13,13,13,0,1,14,13,4,5,6,7,0,1,2,3,4,5,6,7,10,10,2,3,4,5,6,7,0,1,10,10,14,14,10,10,0,1,2,3,4,5,6,7,9,9,9,9,11,9,9,12,0,1,2,3,4,5,6,7,8,8,12,8,12,14,8,8,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,14,11,2,3,4,5,6,7,0,1,11,11,13,15,11,11,0,1,2,3,4,5,6,7,15,12,2,3,12,12,12,12,14,14,12,12,4,5,14,14,0,1,14,14,14,14,6,7,14,9,10,9,9,13,15,9,15,8,13,15,8,8,8,11,15,10,2,3,4,5,6,7,0,1,10,10,10,10,10,10,0,1,2,3,4,5,6,7,13,13,2,3,13,13,13,13,15,15,13,13,4,5,15,15,0,1,15,15,15,15,6,7,0,1,2,3,4,5,6,7,11,11,2,3,4,5,6,7,0,1,11,11,12,11,11,11,8,8,8,8,12,12,8,8,9,11,9,9,9,9,10,9,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,10,10,2,3,4,5,6,7,0,1,10,14,13,10,10,10,12,14,2,3,12,12,12,12,0,1,12,12,4,5,6,7,0,1,2,3,4,5,6,7,8,8,14,13,14,8,8,8,0,1,2,3,4,5,6,7,12,9,9,9,9,14,11,14,0,1,2,3,4,5,6,7,14,14,2,3,4,5,6,7,11,11,2,3,13,13,13,13,13,13,11,11,11,11,11,12,0,1,13,13,4,5,6,7,0,1,2,3,4,5,6,7,10,10,2,3,4,5,6,7,10,8,8,8,8,8,10,10,0,1,2,3,4,5,15,15,0,1,15,15,15,15,9,9,9,9,9,9,9,9,6,7,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7)"
PL/pgSQL function inline_code_block line 16 at EXECUTE statement
DECLARE
	group_info               record;
	node_names               text;
	bucket_str               text;
	query                    text;
	test_case                text;
BEGIN
	test_case := 'create group with buckets not even: some node has no buckets';
	select * from pgxc_group into group_info;
	select string_agg(node_name,',') from pgxc_node, pgxc_group where pgxc_node.oid = any(group_members) and pgxc_group.oid = group_info.oid into node_names;
	
	RAISE INFO 'test %...', test_case;
	select regexp_replace(group_info.group_buckets,'(^|,)0($|,)','\11\2','g') into bucket_str;
	query := 'create node group ' || group_info.group_name || '_dummy with(' || node_names || ') buckets(' || bucket_str || ')';
	execute query;
	
	RAISE INFO 'test % ok', test_case;
	RETURN;
END;
/
INFO:  test create group with buckets not even: some node has no buckets...
ERROR:  Node(0) has no buckets on it.
CONTEXT:  SQL statement "create node group new_group_004_dummy with(datanode1,datanode2,datanode3,datanode4,datanode5,datanode6,datanode7,datanode8,datanode9,datanode10,datanode11,datanode12,datanode13,datanode14,datanode15,datanode16) buckets(14,15,12,14,14,15,8,8,15,13,14,12,13,14,15,9,9,9,9,9,14,9,15,10,15,11,10,10,10,10,10,10,11,11,11,11,11,11,11,11,12,12,11,12,12,12,12,12,13,13,12,12,13,13,13,13,14,14,13,13,13,13,14,14,15,15,14,14,14,14,15,15,1,1,15,15,15,15,15,15,8,8,8,8,10,8,11,8,1,1,2,3,4,5,6,7,13,9,9,9,9,12,9,13,1,1,2,3,4,5,6,7,1,1,2,3,4,5,6,7,1,1,2,3,4,5,6,7,10,10,2,3,4,5,6,7,1,1,12,10,14,10,10,10,11,11,2,3,4,5,6,7,1,1,11,14,11,15,11,11,8,8,12,8,13,8,8,8,1,1,2,3,4,5,6,7,1,1,2,3,4,5,6,7,9,9,14,9,14,11,9,15,1,1,2,3,4,5,6,7,14,12,2,3,15,12,12,12,1,1,12,12,4,5,6,7,15,13,2,3,13,13,13,13,1,1,13,13,4,5,6,7,1,1,2,3,4,5,6,7,10,14,8,11,15,8,8,8,15,10,2,3,4,5,6,7,1,1,10,10,10,10,12,10,14,14,2,3,4,5,14,14,9,9,9,9,9,12,10,9,11,12,14,14,14,14,6,7,1,1,11,11,13,11,11,11,15,15,2,3,4,5,15,15,1,1,15,15,15,15,6,7,1,1,2,3,4,5,6,7,8,8,12,8,13,8,13,8,1,1,2,3,4,5,6,7,1,1,2,3,4,5,6,7,1,1,2,3,4,5,6,7,12,12,2,3,12,12,14,12,9,13,11,9,9,9,9,9,10,10,12,15,4,5,6,7,1,1,10,10,10,10,14,10,1,1,2,3,4,5,6,7,13,13,2,3,14,13,13,15,11,8,8,8,8,8,14,8,11,14,13,13,4,5,6,7,1,1,11,11,15,12,11,11,1,1,2,3,4,5,6,7,1,1,2,3,4,5,6,7,1,1,2,3,4,5,6,7,9,14,10,9,9,15,12,11,1,1,2,3,4,5,6,7,1,1,2,3,4,5,6,7,14,14,2,3,4,5,14,14,15,8,15,12,8,13,8,8,12,10,14,14,14,14,6,7,1,1,10,10,13,10,10,10,1,1,2,3,4,5,6,7,15,15,2,3,4,5,15,15,12,12,15,15,12,12,12,12,11,11,12,12,15,15,6,7,9,9,9,13,9,9,10,9,1,1,11,11,11,11,11,11,1,1,2,3,4,5,6,7,13,8,8,8,10,8,8,8,1,1,2,3,4,5,6,7,13,13,2,3,13,13,13,13,1,1,14,13,4,5,6,7,1,1,2,3,4,5,6,7,10,10,2,3,4,5,6,7,1,1,10,10,14,14,10,10,1,1,2,3,4,5,6,7,9,9,9,9,11,9,9,12,1,1,2,3,4,5,6,7,8,8,12,8,12,14,8,8,1,1,2,3,4,5,6,7,1,1,2,3,4,5,6,7,14,11,2,3,4,5,6,7,1,1,11,11,13,15,11,11,1,1,2,3,4,5,6,7,15,12,2,3,12,12,12,12,14,14,12,12,4,5,14,14,1,1,14,14,14,14,6,7,14,9,10,9,9,13,15,9,15,8,13,15,8,8,8,11,15,10,2,3,4,5,6,7,1,1,10,10,10,10,10,10,1,1,2,3,4,5,6,7,13,13,2,3,13,13,13,13,15,15,13,13,4,5,15,15,1,1,15,15,15,15,6,7,1,1,2,3,4,5,6,7,11,11,2,3,4,5,6,7,1,1,11,11,12,11,11,11,8,8,8,8,12,12,8,8,9,11,9,9,9,9,10,9,1,1,2,3,4,5,6,7,1,1,2,3,4,5,6,7,1,1,2,3,4,5,6,7,10,10,2,3,4,5,6,7,1,1,10,14,13,10,10,10,12,14,2,3,12,12,12,12,1,1,12,12,4,5,6,7,1,1,2,3,4,5,6,7,8,8,14,13,14,8,8,8,1,1,2,3,4,5,6,7,12,9,9,9,9,14,11,14,1,1,2,3,4,5,6,7,14,14,2,3,4,5,6,7,11,11,2,3,13,13,13,13,13,13,11,11,11,11,11,12,1,1,13,13,4,5,6,7,1,1,2,3,4,5,6,7,10,10,2,3,4,5,6,7,10,8,8,8,8,8,10,10,1,1,2,3,4,5,15,15,1,1,15,15,15,15,9,9,9,9,9,9,9,9,6,7,1,1,2,3,4,5,6,7,1,1,2,3,4,5,6,7,1,1,2,3,4,5,6,7,1,1,2,3,4,5,6,7)"
PL/pgSQL function inline_code_block line 14 at EXECUTE statement
DECLARE
	node_count               int;
	group_info               record;
	node_names               text;
	bucket_str               text;
	query                    text;
	test_case                text;
BEGIN
	test_case := 'create group with bucket id out of range';
	select * from pgxc_group into group_info;
	select string_agg(node_name,',') from pgxc_node, pgxc_group where pgxc_node.oid = any(group_members) and pgxc_group.oid = group_info.oid into node_names;
	
	RAISE INFO 'test %...', test_case;
	select count(*) from pgxc_node, pgxc_group where pgxc_node.oid = any(group_members) and pgxc_group.oid = group_info.oid into node_count;
	select regexp_replace(group_info.group_buckets,'(^|,)0($|,)','\1'||node_count||'\2') into bucket_str;
	query := 'create node group ' || group_info.group_name || '_dummy with(' || node_names || ') buckets(' || bucket_str || ')';
	execute query;
	
	RAISE INFO 'test % ok', test_case;
	RETURN;
END;
/
INFO:  test create group with bucket id out of range...
ERROR:  Bucket id(72:16) out of range.
CONTEXT:  SQL statement "create node group new_group_004_dummy with(datanode1,datanode2,datanode3,datanode4,datanode5,datanode6,datanode7,datanode8,datanode9,datanode10,datanode11,datanode12,datanode13,datanode14,datanode15,datanode16) buckets(14,15,12,14,14,15,8,8,15,13,14,12,13,14,15,9,9,9,9,9,14,9,15,10,15,11,10,10,10,10,10,10,11,11,11,11,11,11,11,11,12,12,11,12,12,12,12,12,13,13,12,12,13,13,13,13,14,14,13,13,13,13,14,14,15,15,14,14,14,14,15,15,16,1,15,15,15,15,15,15,8,8,8,8,10,8,11,8,0,1,2,3,4,5,6,7,13,9,9,9,9,12,9,13,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,10,10,2,3,4,5,6,7,0,1,12,10,14,10,10,10,11,11,2,3,4,5,6,7,0,1,11,14,11,15,11,11,8,8,12,8,13,8,8,8,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,9,9,14,9,14,11,9,15,0,1,2,3,4,5,6,7,14,12,2,3,15,12,12,12,0,1,12,12,4,5,6,7,15,13,2,3,13,13,13,13,0,1,13,13,4,5,6,7,0,1,2,3,4,5,6,7,10,14,8,11,15,8,8,8,15,10,2,3,4,5,6,7,0,1,10,10,10,10,12,10,14,14,2,3,4,5,14,14,9,9,9,9,9,12,10,9,11,12,14,14,14,14,6,7,0,1,11,11,13,11,11,11,15,15,2,3,4,5,15,15,0,1,15,15,15,15,6,7,0,1,2,3,4,5,6,7,8,8,12,8,13,8,13,8,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,12,12,2,3,12,12,14,12,9,13,11,9,9,9,9,9,10,10,12,15,4,5,6,7,0,1,10,10,10,10,14,10,0,1,2,3,4,5,6,7,13,13,2,3,14,13,13,15,11,8,8,8,8,8,14,8,11,14,13,13,4,5,6,7,0,1,11,11,15,12,11,11,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,9,14,10,9,9,15,12,11,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,14,14,2,3,4,5,14,14,15,8,15,12,8,13,8,8,12,10,14,14,14,14,6,7,0,1,10,10,13,10,10,10,0,1,2,3,4,5,6,7,15,15,2,3,4,5,15,15,12,12,15,15,12,12,12,12,11,11,12,12,15,15,6,7,9,9,9,13,9,9,10,9,0,1,11,11,11,11,11,11,0,1,2,3,4,5,6,7,13,8,8,8,10,8,8,8,0,1,2,3,4,5,6,7,13,13,2,3,13,13,13,13,0,1,14,13,4,5,6,7,0,1,2,3,4,5,6,7,10,10,2,3,4,5,6,7,0,1,10,10,14,14,10,10,0,1,2,3,4,5,6,7,9,9,9,9,11,9,9,12,0,1,2,3,4,5,6,7,8,8,12,8,12,14,8,8,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,14,11,2,3,4,5,6,7,0,1,11,11,13,15,11,11,0,1,2,3,4,5,6,7,15,12,2,3,12,12,12,12,14,14,12,12,4,5,14,14,0,1,14,14,14,14,6,7,14,9,10,9,9,13,15,9,15,8,13,15,8,8,8,11,15,10,2,3,4,5,6,7,0,1,10,10,10,10,10,10,0,1,2,3,4,5,6,7,13,13,2,3,13,13,13,13,15,15,13,13,4,5,15,15,0,1,15,15,15,15,6,7,0,1,2,3,4,5,6,7,11,11,2,3,4,5,6,7,0,1,11,11,12,11,11,11,8,8,8,8,12,12,8,8,9,11,9,9,9,9,10,9,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,10,10,2,3,4,5,6,7,0,1,10,14,13,10,10,10,12,14,2,3,12,12,12,12,0,1,12,12,4,5,6,7,0,1,2,3,4,5,6,7,8,8,14,13,14,8,8,8,0,1,2,3,4,5,6,7,12,9,9,9,9,14,11,14,0,1,2,3,4,5,6,7,14,14,2,3,4,5,6,7,11,11,2,3,13,13,13,13,13,13,11,11,11,11,11,12,0,1,13,13,4,5,6,7,0,1,2,3,4,5,6,7,10,10,2,3,4,5,6,7,10,8,8,8,8,8,10,10,0,1,2,3,4,5,15,15,0,1,15,15,15,15,9,9,9,9,9,9,9,9,6,7,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7,0,1,2,3,4,5,6,7)"
PL/pgSQL function inline_code_block line 16 at EXECUTE statement
